internal/entitycache/cache.go

```diff
@@ -0,0 +1,669 @@
+			}
+			return e, nil
+		}
+		if s.err != nil {
+			return nil, errgo.Notef(s.err, "cannot fetch %q", id)
+		}
+		if hasEntry {
+			// The entity is already being fetched. Wait for the fetch
+			// to complete and try again.
+			s.changed.Wait()
+			e, hasEntry = s.entities[*id]
+			continue
+		}
+		// Fetch synchronously (any other goroutines will be
+		// notified when we've retrieved the entity). After the
+		// fetch has completed, the entry in the cache will either
+		// be set to the retrieved entity, or deleted (if the
+		// selected fields have changed).
+		s.entities[*id] = nil
+		version := s.version
+		fields := s.fields
+		s.mu.Unlock()
+		e = s.fetch(id, fields, version, false)
+		s.mu.Lock()
+		// Invariant (from fetch): e != nil || s.err != nil
+	}
+}
+
+// addFields adds the given fields to those that will be fetched
+// when an entity is fetched.
+//
+// Called with s.mu locked.
+func (s *stash) addFields(fields []string) {
+	changed := false
+	for _, field := range fields {
+		if _, ok := s.fields[field]; !ok {
+			changed = true
+			break
+		}
+	}
+	if !changed {
+		return
+	}
+	if len(s.entities) > 0 {
+		// The fields have changed, invalidating our current
+		// cache, so delete all entries.
+		s.entities = make(map[charm.URL]stashEntity)
+		s.version++
+		// There may be several goroutines waiting for pending
+		// entities. Notify them so that they can start a new
+		// fetch.
+		s.changed.Broadcast()
+	}
+	newFields := make(map[string]int)
+	for field := range s.fields {
+		newFields[field] = 1
+	}
+	for _, field := range fields {
+		newFields[field] = 1
+	}
+	s.fields = newFields
+}
+
+// startFetch starts an asynchronous fetch for the given id.
+// If a fetch is already in progress, it does nothing.
+//
+// Called with s.mu locked.
+func (s *stash) startFetch(id *charm.URL) {
+	if _, ok := s.entities[*id]; ok {
+		return
+	}
+	s.entities[*id] = nil
+	// Note that it's only OK to pass s.fields here because
+	// it's never mutated, only replaced.
+	s.wg.Add(1)
+	go s.fetch(id, s.fields, s.version, true)
+}
+
+// fetch fetches the entity with the given id, including the given
+// fields, adds it to the stash and notifies any waiters that the stash
+// has changed. If async is true, s.wg.Done will be called just
+// before fetch returns (this must only be done if s.wg.Add has
+// been called appropriately and fetch is running in its own goroutine).
+//
+// The given entity version holds the version at the time the
+// fetch was started. If the entity version has changed when the result is received,
+// the result is discarded.
+//
+// fetch returns the entity as it would be stored in the cache (notFoundEntity
+// implies not found). It returns nil if and only if some other kind of error has
+// been encountered (in this case the error will be stored in s.err).
+//
+// Called with no locks held.
+func (s *stash) fetch(url *charm.URL, fields map[string]int, version int, async bool) stashEntity {
```

Comment:
Yeah, we debated that. fetchAsync seems like a reasonable solution. I'll see how that feels.